src/rgw/rgw_dedup_table.h

```diff
@@ -0,0 +1,136 @@
+  } __attribute__((__packed__));
+
+  class dedup_table_t {
+  public:
+    // 10 Bytes Value
+    struct value_t {
+      value_t() {
+        this->block_idx = 0xFFFFFFFF;
+        this->count  = 0;
+        this->pad_16 = 0;
+        this->rec_id = 0xFF;
+        this->flags.clear();
+
+      }
+
+      value_t(disk_block_id_t block_id, record_id_t rec_id, bool shared_manifest, bool valid_sha256) {
+        this->block_idx = block_id;
+        this->count  = 1;
+        this->pad_16 = 0;
+        this->rec_id = rec_id;
+        this->flags.clear();
+        this->flags.set_singleton_occupied();
+        if (shared_manifest) {
+          flags.set_shared_manifest();
+        }
+        if (valid_sha256) {
+          flags.set_valid_sha256();
+        }
+      }
+
+      inline void clear_flags() { flags.clear(); }
+      inline bool has_valid_sha256() const { return flags.has_valid_sha256(); }
+      inline void set_valid_sha256() { this->flags.set_valid_sha256(); }
+      inline bool has_shared_manifest() const {return flags.has_shared_manifest(); }
+      inline void set_shared_manifest() { this->flags.set_shared_manifest(); }
+      inline bool is_singleton() const { return flags.is_singleton(); }
+      inline void clear_singleton() { this->flags.clear_singleton(); }
+      inline bool is_occupied() const { return flags.is_occupied(); }
+      inline void set_occupied() { this->flags.set_occupied();  }
+      inline void clear_occupied() { this->flags.clear_occupied(); }
+      inline void set_singleton_occupied() { this->flags.set_singleton_occupied(); }
+
+      disk_block_id_t block_idx;
+      uint16_t        count;
+      uint16_t        pad_16;
+      record_id_t     rec_id;   // TBD use 16bits rec_id
+      dedup_flags_t   flags;
+    } __attribute__((__packed__));
+
+    dedup_table_t(const DoutPrefixProvider* _dpp,
+                  uint32_t _head_object_size,
+                  uint8_t *p_slab,
+                  uint64_t slab_size);
+    int add_entry(key_t *p_key, disk_block_id_t block_id, record_id_t rec_id,
+                  bool shared_manifest, bool has_sha256);
+    int update_entry(key_t *p_key, disk_block_id_t block_id, record_id_t rec_id,
+                     bool shared_manifest, bool valid_sha256);
+
+    int  get_val(const key_t *p_key, struct value_t *p_val /*OUT*/);
+
+    int set_shared_manifest_mode(const key_t *p_key,
+                                 disk_block_id_t block_id,
+                                 record_id_t rec_id);
+    void count_duplicates(uint64_t *p_singleton_count,
+                          uint64_t *p_unique_count,
+                          uint64_t *p_duplicate_count,
+                          uint64_t *p_duplicate_bytes_approx);
+    void remove_singletons_and_redistribute_keys();
+  private:
+    // 32 Bytes unified entries
+    struct table_entry_t {
+      key_t key;
+      value_t val;
+    } __attribute__((__packed__));
+
+    uint32_t find_entry(const key_t *p_key);
```

Comment:
nit: please add `const` to function definitions when applicable (help with readability)